rules/55-zz-scsi_id.rules

```diff
@@ -0,0 +1,20 @@
+ACTION=="remove", GOTO="scsi_id_end"
+ENV{ID_SERIAL}=="?*", GOTO="scsi_id_end"
+
+# SCSI devices are normally handled in 55-sg3_utils.rules.
+# Add fallback rules here for users who don't have sg3_utils.
+# Call after other 55- rules, but before 56- (multipath!)
+# This may go away once scsi_id is removed from udev
+
+KERNEL=="sd*[!0-9]|sr*", \
+	IMPORT{program}="scsi_id --export --whitelisted -d $devnode", \
+        ENV{.ID_BUS}="scsi"
+KERNEL=="cciss*", ENV{DEVTYPE}=="disk", \
+	IMPORT{program}="scsi_id --export --whitelisted -d $devnode", \
+        ENV{.ID_BUS}="cciss"
+
+# This is executed after other rules that may have set ID_BUS already
+# (e.g. USB). Don't overwrite it.
+ENV{ID_BUS}!="?*", ENV{.ID_BUS}=="?*", ENV{ID_BUS}="$env{.ID_BUS}"
+
```

Comment:
Couldn't the value of `ENV{ID_BUS}` be tested at the beginning of this rule file and skip all calls to scsi_id if it's already set ?

This way you a) don't need an temporary variable ".ID_BUS" (the '.' prefix looks a bit weird BTW) b) don't call scsi_id needlessly if ID_BUS is already initialized.

Reply:
> Couldn't the value of `ENV{ID_BUS}` be tested at the beginning of this rule file and skip all calls to scsi_id if it's already set ?

No. This code has to be executed if and only if `ID_SERIAL` is unset yet (likely because **sg3_utils**) was not installed). Testing `ID_BUS` equivalent to testing `ID_SERIAL`. `scsi_id --export` sets lots of environment variables, including  `ID_SERIAL`.
 
> This way you a) don't need an temporary variable ".ID_BUS" (the '.' prefix looks a bit weird BTW) b) don't call scsi_id needlessly if ID_BUS is already initialized.

Well, it is the syntax that **udev** mandates for temporary variables. See **udev(7)**:

>  Property names with a leading "."  are neither stored in the database nor exported to events or external tools (run by, for example, the PROGRAM match key)

Reply:
OK, thanks.